src/main/java/frc/robot/subsystems/Vision/VisionIO.java

```diff
@@ -4,83 +4,142 @@
 
-  public abstract List<VisionMeasurement> getMeasurement(Pose2d previousPose);
+        public void getLoggedPhotonTrackedTarget(PhotonTrackedTarget target, LogTable table, String name) {
+            double[] translation = table.getDoubleArray("translation " + name, new double[3]);
+            double[] rotation = table.getDoubleArray("rotation " + name, new double[4]);
+            double[] altTranslation = table.getDoubleArray("translation alt " + name, new double[3]);
+            double[] altRotation = table.getDoubleArray("rotation alt " + name, new double[4]);
+            double[] detectedCornersX = table.getDoubleArray("detected corners x "  + name, new double[4]);
+            double[] detectedCornersY = table.getDoubleArray("detected corners y "  + name, new double[4]);
+            double[] minAreaRectCornersX = table.getDoubleArray("min area rect corners x " + name, new double[4]);
+            double[] minAreaRectCornersY = table.getDoubleArray("min area rect corners y " + name, new double[4]);
 
-  public void updateInputs(VisionIOInputs inputs, Pose3d robotPose);
+            List<TargetCorner> detectedCorners = new ArrayList<>();
+            List<TargetCorner> minAreaRectCorners = new ArrayList<>();
+
+            for (int i = 0; i <4; i++) {
+                detectedCorners.add(new TargetCorner(detectedCornersX[i], detectedCornersY[i]));
+                minAreaRectCorners.add(new TargetCorner(minAreaRectCornersX[i], minAreaRectCornersY[i]));
+            }
+            Transform3d pose = getLoggedTransform3d(translation, rotation);
+            Transform3d altPose = getLoggedTransform3d(altTranslation, altRotation);
+            targets.add(
+                new PhotonTrackedTarget(
+                    table.getDouble("yaw " + name, target.getYaw()),
+                    table.getDouble("pitch " + name, target.getPitch()),
+                    table.getDouble("area " + name, target.getArea()),
+                    table.getDouble("skew " + name, target.getSkew()),
+                    (int)(table.getInteger("fiducial id " + name, target.getFiducialId())),
+                    pose,
+                    altPose,
+                    table.getDouble("pose ambiguity " + name, target.getPoseAmbiguity()),
+                    minAreaRectCorners,
+                    detectedCorners)
+            );
+        }
+
+        @Override
+        public void toLog(LogTable table) {
+            table.put("timestamp", timestamp);
+            table.put("latency", timeSinceLastTimestamp);
+            for (int i = 0; i < targets.size(); i++) {
+                logPhotonTrackedTarget(targets.get(i), table, String.valueOf(i));
+            }
+        }
+        
+        @Override
+        public void fromLog(LogTable table) {
+            timestamp = table.getDouble("timestamp", timestamp);
+            timeSinceLastTimestamp = table.getDouble("latency", timeSinceLastTimestamp);
+            for (int i = 0; i < targets.size(); i++) {
```

Comment:
keep in mind that targets is an empty list at this point, so youll need to get the size from something else. probably should add `numTags` or something to `table`